### PR TITLE
[YUNIKORN-1155] Placeholder timeout should not be initialized to the default value during recovery

### PR DESCRIPTION
**Pull request by pbacsko**
### What is this PR for?
When Yunikorn is restarted, the placeholder timeout is reset to the original value, regardless of how much time the placeholder had already spent in pending/running state. This PR fixes this issue by checking the creationTime attribute of the pod and adjusting it if necessary.

### What type of PR is it?
* [x] - Bug Fix
* [ ] - Improvement
* [ ] - Feature
* [ ] - Documentation
* [ ] - Hot Fix
* [ ] - Refactoring

### Todos
* [ ] - Task

### What is the Jira issue?
https://issues.apache.org/jira/browse/YUNIKORN-1155

### Questions:
* [ ] - The licenses files need update.
* [ ] - There is breaking changes for older versions.
* [ ] - It needs documentation.

### CODE REVIEW

File: pkg/scheduler/objects/application.go [hunk 1]
```diff
@@ -114,10 +115,19 @@ func NewApplication(siApp *si.AddApplicationRequest, ugi security.UserGroup, eve
...
 	placeholderTimeout := common.ConvertSITimeout(siApp.ExecutionTimeoutMilliSeconds)
 	if time.Duration(0) == placeholderTimeout {
 		placeholderTimeout = defaultPlaceholderTimeout
 	}
+
+	var timeoutReached bool
+	timeoutReached, placeholderTimeout = app.adjustPlaceholderTimeout(placeholderTimeout, siApp)
+	if timeoutReached {
+		// Yunikorn was restarted and the placeholder timed out in the meantime
+		placeholderTimeout = time.Millisecond
```
Comment (pbacsko) [inline pkg/scheduler/objects/application.go:128]: 1 msec is basically instant timeout. There is a guard against the value 0 somewhere else, so this is the smallest allowed value.

File: pkg/scheduler/objects/application.go [hunk 2]
```diff
@@ -137,6 +147,37 @@ func NewApplication(siApp *si.AddApplicationRequest, ugi security.UserGroup, eve
 	return app
 }
 
+// Adjust the placeholder timeout if "creationTime" is defined. It's used during Yunikorn restart,
+// in order to properly track how long a placeholder pod should be in "Running" state.
+func (sa *Application) adjustPlaceholderTimeout(placeholderTimeout time.Duration, siApp *si.AddApplicationRequest) (bool, time.Duration) {
+	creationTimeTag := siApp.Tags["creationTime"]
```
Comment (pbacsko) [inline pkg/scheduler/objects/application.go:153]: As I mentioned in https://github.com/apache/yunikorn-k8shim/pull/405/, this is a tag, but `AddApplicationRequest` can be extended with a `creationTime` field.

File: pkg/scheduler/objects/application.go [hunk 3]
```diff
@@ -137,6 +147,37 @@ func NewApplication(siApp *si.AddApplicationRequest, ugi security.UserGroup, eve
...
+	var creationTime int64
+	if creationTimeTag != "" {
+		var err error
+		creationTime, err = strconv.ParseInt(creationTimeTag, 10, 64)
+		if err != nil {
```
Comment (wilfred-s) [inline pkg/scheduler/objects/application.go:158]: Move to `common.ConvertSITimestamp(ts string) time.Time` for the conversion.

Does this work as a simpler option to do the calculation:
```
created := common.ConvertSITimestamp(creationTimeTag)
newPT := created.Add(placeholderTimeout).sub(time.Now())
```

**Review by wilfred-s — CHANGES_REQUESTED** [review]

File: pkg/common/utils.go [hunk 4]
```diff
@@ -115,6 +120,54 @@ func ConvertSITimeout(millis int64) time.Duration {
 	return time.Duration(result)
 }
 
+// ConvertSITimeoutWithAdjustment Similar to ConvertSITimeout, but this function also adjusts the timeout if
```
Comment (craigcondit) [inline pkg/common/utils.go:123]: Can we get unit tests for these please?

**Review by craigcondit — CHANGES_REQUESTED** [review]
Looks good, but let's get some unit tests around these new functions.

**Review by craigcondit — APPROVED** [review]
+1 LGTM.

File: pkg/common/utils.go [hunk 5]
```diff
@@ -30,6 +30,14 @@ import (
 
 	"github.com/apache/yunikorn-core/pkg/log"
 	interfaceCommon "github.com/apache/yunikorn-scheduler-interface/lib/go/common"
+	"github.com/apache/yunikorn-scheduler-interface/lib/go/si"
+)
+
+const CreationTime = "CreationTime"
```
Comment (wilfred-s) [inline pkg/common/utils.go:36]: This should be moved to the SI as it used in both shim and core. Needs to be added to YUNIKORN-1122 like you did with `ready`.

File: pkg/common/utils.go [hunk 6]
```diff
@@ -30,6 +30,14 @@ import (
...
+
+var (
+	currentTime = time.Now
+	Undefined   = time.Time{}
```
Comment (wilfred-s) [inline pkg/common/utils.go:40]: I don't think we need these, see other comments.

File: pkg/common/utils.go [hunk 7]
```diff
@@ -115,6 +123,56 @@ func ConvertSITimeout(millis int64) time.Duration {
...
+}
+
+func adjustTimeout(timeout time.Duration, siApp *si.AddApplicationRequest) time.Duration {
+	creationTimeTag := siApp.Tags[interfaceCommon.DomainYuniKorn+CreationTime]
+	if creationTimeTag == "" {
+		return timeout
+	}
+
+	created := ConvertSITimestamp(creationTimeTag)
+	if created == Undefined {
```
Comment (wilfred-s) [inline pkg/common/utils.go:143]: use `created.IsZero()` for this instead of using a variable.

File: pkg/common/utils.go [hunk 8]
```diff
@@ -115,6 +123,56 @@ func ConvertSITimeout(millis int64) time.Duration {
...
+		return timeout
+	}
+	now := currentTime()
```
Comment (wilfred-s) [inline pkg/common/utils.go:146]: Testing and all other cases should never need to override `time.Now()`. We do not care if it is exactly 1 or 10 minutes as long as it is less than what is set as the timeout.

File: pkg/common/utils.go [hunk 9]
```diff
@@ -115,6 +123,56 @@ func ConvertSITimeout(millis int64) time.Duration {
 	return time.Duration(result)
 }
 
+// ConvertSITimeoutWithAdjustment Similar to ConvertSITimeout, but this function also adjusts the timeout if
+// "creationTime" is defined. It's used during Yunikorn restart, in order to properly track how long a placeholder pod should
+// be in "Running" state.
+func ConvertSITimeoutWithAdjustment(siApp *si.AddApplicationRequest) time.Duration {
+	result := ConvertSITimeout(siApp.ExecutionTimeoutMilliSeconds)
```
Comment (wilfred-s) [inline pkg/common/utils.go:130]: What happens if I do not have a timeout set? Setting the default is part of the application code which means we'll break at this point. Setting the default should be moved into the `ConvertSITimeout()` call.

File: pkg/common/utils_test.go [hunk 10]
```diff
@@ -133,3 +135,55 @@ func TestGetRequiredNodeFromAsk(t *testing.T) {
 	nodeName = GetRequiredNodeFromTag(tag)
 	assert.Equal(t, nodeName, "Node2")
 }
+
+func TestConvertSITimeoutWithAdjustment(t *testing.T) {
```
Comment (wilfred-s) [inline pkg/common/utils_test.go:139]: Missing test cases with no timeout set and using the default.

File: pkg/common/utils.go [hunk 11]
```diff
@@ -115,6 +123,56 @@ func ConvertSITimeout(millis int64) time.Duration {
...
+
+	log.Logger().Info("Adjusting placeholder timeout",
+		zap.Duration("timeout duration", timeout),
+		zap.Time("creation time", created),
+		zap.Time("expected timeout", expectedTimeout))
+
+	return expectedTimeout.Sub(now)
+}
+
+func ConvertSITimestamp(ts string) time.Time {
```
Comment (wilfred-s) [inline pkg/common/utils.go:165]: Design decision to be made: do we support negative numbers in the timestamp? i.e. dates before `1970-01-01 00:00:00`
Also not sure if you want to print a warning when the string passed in is empty. Pollutes the logs and would happen often.

**Comment by wilfred-s** [comment]
Looking through the change the default timeout handling is broken for this change. Checking if the default timeout needs to be used is done after the calculation. Which means that old placeholders using the default timeout do not get adjusted.

File: pkg/common/utils.go (hunk 11 above)
Comment (pbacsko) [reply pkg/common/utils.go:161]: Fixed

File: pkg/common/utils.go (hunk 9 above)
Comment (pbacsko) [reply pkg/common/utils.go:123]: Fixed.

File: pkg/common/utils.go (hunk 8 above)
Comment (pbacsko) [reply pkg/common/utils.go]: Fixed

**Review by wilfred-s — APPROVED** [review]
LGTM defaults work again

**Pull request closed without merging**